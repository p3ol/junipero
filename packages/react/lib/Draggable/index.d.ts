import { DragEvent, ComponentPropsWithRef } from 'react';
declare interface DraggableProps extends ComponentPropsWithRef<any> {
    children?: JSX.Element;
    className?: string;
    data?: any;
    disabled?: boolean;
    dragImage?: Element;
    dragImageOffset?: {
        x: number;
        y: number;
    };
    onDrag?(e: Event): void;
    onBeforeDragStart?(e: DragEvent): void;
    onDragStart?(e: DragEvent, target?: DOMRect): void;
    onDragEnd?(e: DragEvent): void;
}
declare const Draggable: import("react").ForwardRefExoticComponent<Omit<DraggableProps, "ref"> & import("react").RefAttributes<any>>;
export default Draggable;
