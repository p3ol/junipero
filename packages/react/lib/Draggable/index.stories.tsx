import { useState } from 'react';

import Card from '../Card';
import Droppable from '../Droppable';
import Draggable from '.';

export default { title: 'react/Draggable' };

const Item = ({ name }: { name: string}) => (
  <Draggable data={{ name }}>
    <Card>{ name }</Card>
  </Draggable>
);

export const Basic = () => {
  const [items, setItems] = useState([]);

  return (
    <div style={{ display: 'flex' }}>
      <div style={{ marginRight: 20 }}>
        { ['Item 1', 'Item 2', 'Item 3'].map((name, i) => (
          <Item key={i} name={name} />
        )) }
      </div>
      <Droppable onDrop={(data: any) => setItems([...items, data])}>
        <div style={{ padding: 20, border: '1px solid #ccc' }}>
          { items.map((item, i) => <div key={i}>{ item.name }</div>) }
        </div>
      </Droppable>
    </div>
  );
};
