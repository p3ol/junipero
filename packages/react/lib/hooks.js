import { useContext } from 'react';

import { FieldControlContext } from './contexts';

export const useFieldControl = () => useContext(FieldControlContext);
