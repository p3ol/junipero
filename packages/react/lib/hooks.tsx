import { useContext } from 'react';

import {
  AlertsContextType,
  AlertsContext,
  DropdownContext,
  DropdownContextType,
  FieldContextType,
  FieldControlContext,
  ListContext,
  ListContextType,
  ToastsContext,
  ToastsContextType,
  ModalContext,
  ModalContextType,
} from './contexts';

export const useFieldControl =
  () => useContext<FieldContextType>(FieldControlContext);
export const useDropdown = () =>
  useContext<DropdownContextType>(DropdownContext);
export const useAlerts = () => useContext<AlertsContextType>(AlertsContext);
export const useList = () => useContext<ListContextType>(ListContext);
export const useToasts = () => useContext<ToastsContextType>(ToastsContext);
export const useModal = () => useContext<ModalContextType>(ModalContext);

export { AlertsContext, DropdownContext, FieldControlContext, ListContext, ToastsContext, ModalContext };
