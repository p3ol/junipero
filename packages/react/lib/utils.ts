import type { StateContent } from './types';

export const mockState = <T extends StateContent>(
  state: T,
  action : ((prev: T) => T) | Partial<T>
) => typeof action === 'function'
    ? action(state) : ({ ...state, ...action as T });
