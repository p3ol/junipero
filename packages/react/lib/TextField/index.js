import {
  forwardRef,
  useEffect,
  useImperativeHandle,
  useReducer,
  useRef,
} from 'react';
import { classNames, mockState, exists } from '@junipero/core';
import PropTypes from 'prop-types';

import { useFieldControl } from '../hooks';

const TextField = forwardRef(({
  min,
  max,
  autoFocus,
  children,
  className,
  valid,
  value,
  disabled = false,
  required = false,
  type = 'text',
  onBlur,
  onChange,
  onFocus,
  onWheel,
  onValidate = (val, { required }) => !!val || !required,
  ...rest
}, ref) => {
  const innerRef = useRef();
  const inputRef = useRef();
  const { update: updateControl } = useFieldControl();
  const [state, dispatch] = useReducer(mockState, {
    value: value ?? '',
    valid: valid ?? false,
    focused: autoFocus ?? false,
    dirty: false,
  });

  useImperativeHandle(ref, () => ({
    innerRef,
    inputRef,
    value: state.value,
    valid: state.valid,
    dirty: state.dirty,
    focused: state.focused,
    focus,
    blur,
    reset,
    setDirty,
    isJunipero: true,
  }));

  useEffect(() => {
    if (exists(value)) {
      state.value = value;
      state.valid = onValidate(state.value, { dirty: state.dirty, required });
      dispatch({ value: state.value, valid: state.valid });
      updateControl?.({ valid: state.valid, dirty: state.dirty });
    }
  }, [value]);

  useEffect(() => {
    dispatch({ valid: valid ?? false });
    updateControl?.({ valid: state.valid });
  }, [valid]);

  const parseValue = value => {
    const value_ = value.replace(',', '.');
    const parsedValue = parseFloat(value_);

    return !isNaN(parsedValue) ? parsedValue : value_;
  };

  const onChange_ = e => {
    /* html input disabled attribute will prevent onChange if present anyway */
    /* istanbul ignore next: just in case */
    if (disabled) {
      return;
    }

    if (type === 'number') {
      value = parseValue(e?.target?.value);
    } else {
      value = e?.target?.value;
    }

    state.value = value;
    state.dirty = true;
    state.valid = onValidate(
      state.value, { dirty: state.dirty, required }
    );
    dispatch({ value: state.value, dirty: state.dirty });
    onChange?.({ value: state.value, valid: state.valid, dirty: state.dirty });
    updateControl?.({ dirty: true, valid: state.valid });
  };

  const onFocus_ = e => {
    dispatch({ focused: true });
    updateControl?.({ focused: true });
    onFocus?.(e);
  };

  const onBlur_ = e => {
    dispatch({ focused: false });
    updateControl?.({ focused: false });
    onBlur?.(e);
  };

  const focus = () => {
    inputRef.current?.focus();
  };

  const blur = () => {
    inputRef.current?.blur();
  };

  const reset = () => {
    dispatch({ value: value ?? '', valid: valid ?? false, dirty: false });
    updateControl?.({ dirty: false, valid: valid ?? false });
  };

  const isEmpty = () =>
    !exists(state.value) || state.value === '';

  const setDirty = dirty => {
    state.dirty = dirty;
    dispatch({ dirty });
    updateControl?.({ dirty });
  };

  const Tag = rest.rows > 1 ? 'textarea' : 'input';

  return (
    <div
      ref={innerRef}
      className={classNames(
        'junipero',
        'text-field',
        state.dirty ? 'dirty' : 'pristine',
        !state.valid && state.dirty ? 'invalid' : 'valid',
        { empty: isEmpty(), focused: state.focused, disabled },
        className
      )}
    >
      <Tag
        { ...rest }
        min={min}
        max={max}
        autoFocus={autoFocus}
        className="field"
        ref={inputRef}
        disabled={disabled}
        required={required}
        type={Tag === 'textarea' ? undefined : type}
        value={state.value}
        onChange={onChange_}
        onFocus={onFocus_}
        onBlur={onBlur_}
        onWheel={onWheel ?? (() => inputRef.current?.blur())}
      />
      { children }
    </div>
  );
});

TextField.displayName = 'TextField';
TextField.propTypes = {
  min: PropTypes.number,
  max: PropTypes.number,
  autoFocus: PropTypes.bool,
  disabled: PropTypes.bool,
  required: PropTypes.bool,
  type: PropTypes.string,
  valid: PropTypes.bool,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  onBlur: PropTypes.func,
  onChange: PropTypes.func,
  onFocus: PropTypes.func,
  onWheel: PropTypes.func,
  onValidate: PropTypes.func,
};

export default TextField;
