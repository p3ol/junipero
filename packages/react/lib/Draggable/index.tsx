import {
  Children,
  cloneElement,
  forwardRef,
  useState,
  DragEvent,
  ComponentPropsWithRef,
  ReactNode,
} from 'react';
import { classNames } from '@junipero/core';
import { useTimeout } from '@junipero/hooks';
import PropTypes from 'prop-types';

import { ForwardedProps } from '../utils';

declare interface DraggableProps extends ComponentPropsWithRef<any> {
  children?: JSX.Element;
  className?: string;
  data?: any;
  disabled?: boolean;
  dragImage?: Element;
  dragImageOffset?: { x: number; y: number };
  onDrag?(e: Event): void;
  onBeforeDragStart?(e: DragEvent): void;
  onDragStart?(e: DragEvent, target?: DOMRect): void;
  onDragEnd?(e: DragEvent): void;
}

const Draggable = forwardRef(({
  className,
  children,
  dragImage,
  disabled = false,
  data = {},
  dragImageOffset = { x: 0, y: 0 },
  onDrag,
  onBeforeDragStart,
  onDragStart,
  onDragEnd,
  ...rest
}: DraggableProps, ref) => {
  const [dragged, setDragged] = useState(false);
  const [dragAnimation, setDragAnimation] = useState(false);

  useTimeout(() => {
    if (dragAnimation) {
      setDragged(true);
      setDragAnimation(false);
    }
  }, 0, [dragAnimation]);

  const onDragStart_ = (e: DragEvent<HTMLElement>) => {
    if (disabled) {
      return;
    }

    onBeforeDragStart?.(e);

    const targetRect = e.currentTarget.getBoundingClientRect();

    try {
      e.dataTransfer.setData?.('text', JSON.stringify(data));

      if (dragImage) {
        e.dataTransfer.setDragImage(
          dragImage,
          dragImageOffset.x,
          dragImageOffset.y
        );
      }
    } catch (err) {}

    setDragAnimation(true);
    onDragStart?.(e, targetRect);
  };

  const onDragEnd_ = e => {
    if (disabled) {
      return;
    }

    setDragged(false);
    onDragEnd?.(e);
  };

  const onDrag_ = e => {
    if (disabled) {
      return;
    }

    onDrag?.(e);
  };

  const child = Children.only(children);

  return cloneElement(child, {
    ...rest,
    ref,
    className: classNames(
      className,
      child.props.className,
      {
        dragging: !disabled && dragAnimation,
        dragged: !disabled && dragged,
        draggable: !disabled,
      }
    ),
    draggable: !disabled,
    onDragStart: onDragStart_,
    onDrag: onDrag_,
    onDragEnd: onDragEnd_,
  });
}) as ForwardedProps<DraggableProps, any>;

Draggable.displayName = 'Draggable';
Draggable.propTypes = {
  disabled: PropTypes.bool,
  data: PropTypes.oneOfType([
    PropTypes.object,
    PropTypes.array,
    PropTypes.string,
    PropTypes.number,
  ]),
  dragImage: PropTypes.any, //TODO fixme
  dragImageOffset: PropTypes.any, //TODO fixme
  onDrag: PropTypes.func,
  onBeforeDragStart: PropTypes.func,
  onDragStart: PropTypes.func,
  onDragEnd: PropTypes.func,
};

export default Draggable;
