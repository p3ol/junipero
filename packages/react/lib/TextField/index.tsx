import {
  type ChangeEvent,
  type FocusEvent,
  type RefObject,
  type WheelEvent,
  useEffect,
  useImperativeHandle,
  useReducer,
  useRef,
} from 'react';
import { classNames, exists } from '@junipero/core';

import type {
  FieldContent,
  JuniperoRef,
  SpecialComponentPropsWithRef,
} from '../types';
import { mockState } from '../utils';
import { useFieldControl } from '../hooks';

export declare interface TextFieldRef extends JuniperoRef {
  dirty: boolean;
  focused: boolean;
  valid: boolean;
  value: string | number;
  blur(): void;
  focus(): void;
  reset(): void;
  setDirty(dirty: boolean): void;
  innerRef: RefObject<HTMLDivElement>;
  inputRef: RefObject<HTMLInputElement | HTMLTextAreaElement>;
}

export declare interface TextFieldProps
  extends SpecialComponentPropsWithRef<any, TextFieldRef> {
  autoFocus?: boolean;
  disabled?: boolean;
  required?: boolean;
  type?: string;
  valid?: boolean;
  value?: string | number;
  onBlur?(event: FocusEvent<HTMLInputElement | HTMLTextAreaElement>): void;
  onChange?(field: FieldContent<string | number>): void;
  onFocus?(event: FocusEvent<HTMLInputElement | HTMLTextAreaElement>): void;
  onWheel?(event: WheelEvent<HTMLInputElement | HTMLTextAreaElement>): void;
  onValidate?(
    val: string | number,
    flags: { required: boolean; dirty: boolean }
  ): boolean;
}

export declare interface TextFieldState {
  value: string | number;
  valid: boolean;
  dirty: boolean;
  focused: boolean;
}

const TextField = ({
  ref,
  autoFocus,
  children,
  className,
  valid,
  value,
  disabled = false,
  required = false,
  type = 'text',
  onBlur,
  onChange,
  onFocus,
  onWheel,
  onValidate = (val: string | number, opts?: { required: boolean }) =>
    !!val || !opts?.required,
  ...rest
}: TextFieldProps) => {
  const innerRef = useRef<HTMLDivElement>(null);
  const inputRef = useRef<HTMLInputElement & HTMLTextAreaElement>(null);
  const { update: updateControl } = useFieldControl();
  const [state, dispatch] = useReducer(mockState<TextFieldState>, {
    value: value ?? '',
    valid: valid ?? false,
    focused: autoFocus ?? false,
    dirty: false,
  });

  useImperativeHandle(ref, () => ({
    innerRef,
    inputRef,
    value: state.value,
    valid: state.valid,
    dirty: state.dirty,
    focused: state.focused,
    focus,
    blur,
    reset,
    setDirty,
    isJunipero: true,
  }));

  useEffect(() => {
    if (exists(value)) {
      state.value = value;
      state.valid = onValidate(state.value, { dirty: state.dirty, required });
      dispatch({ value: state.value, valid: state.valid });
      updateControl?.({ valid: state.valid, dirty: state.dirty });
    }
  }, [value]);

  useEffect(() => {
    dispatch({ valid: valid ?? false });
    updateControl?.({ valid: state.valid });
  }, [valid]);

  const parseValue = (value?: string) => {
    const value_ = value.replace(',', '.');
    const parsedValue = parseFloat(value_);

    return !isNaN(parsedValue) ? parsedValue : value_;
  };

  const onChange_ = (
    e: ChangeEvent<HTMLTextAreaElement | HTMLInputElement>
  ) => {
    /* html input disabled attribute will prevent onChange if present anyway */
    /* istanbul ignore next: just in case */
    if (disabled) {
      return;
    }

    if (type === 'number') {
      value = parseValue(e?.target?.value);
    } else {
      value = e?.target?.value;
    }

    state.value = value;
    state.dirty = true;
    state.valid = onValidate(
      state.value, { dirty: state.dirty, required }
    );
    dispatch({ value: state.value, dirty: state.dirty });
    onChange?.({ value: state.value, valid: state.valid, dirty: state.dirty });
    updateControl?.({ dirty: true, valid: state.valid });
  };

  const onFocus_ = (e: FocusEvent<HTMLInputElement | HTMLTextAreaElement>) => {
    dispatch({ focused: true });
    updateControl?.({ focused: true });
    onFocus?.(e);
  };

  const onBlur_ = (e: FocusEvent<HTMLInputElement | HTMLTextAreaElement>) => {
    dispatch({ focused: false });
    updateControl?.({ focused: false });
    onBlur?.(e);
  };

  const focus = () => {
    inputRef.current?.focus();
  };

  const blur = () => {
    inputRef.current?.blur();
  };

  const reset = () => {
    dispatch({ value: value ?? '', valid: valid ?? false, dirty: false });
    updateControl?.({ dirty: false, valid: valid ?? false });
  };

  const isEmpty = () =>
    !exists(state.value) || state.value === '';

  const setDirty = (dirty: boolean) => {
    state.dirty = dirty;
    dispatch({ dirty });
    updateControl?.({ dirty });
  };

  const Tag = rest.rows > 1 ? 'textarea' : 'input';

  return (
    <div
      ref={innerRef}
      className={classNames(
        'junipero',
        'text-field',
        state.dirty ? 'dirty' : 'pristine',
        !state.valid && state.dirty ? 'invalid' : 'valid',
        { empty: isEmpty(), focused: state.focused, disabled },
        className
      )}
    >
      <Tag
        { ...rest }
        autoFocus={autoFocus}
        className="field"
        ref={inputRef}
        disabled={disabled}
        required={required}
        type={Tag === 'textarea' ? undefined : type}
        value={state.value}
        onChange={onChange_}
        onFocus={onFocus_}
        onBlur={onBlur_}
        onWheel={onWheel ?? ((_: WheelEvent) => inputRef.current?.blur())}
      />
      { children }
    </div>
  );
};

TextField.displayName = 'TextField';

export default TextField;
