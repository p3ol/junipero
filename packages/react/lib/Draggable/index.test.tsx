import { fireEvent, render } from '@testing-library/react';
import { vi } from 'vitest';

import { sleep } from '../../tests/utils';
import Draggable from './index';

describe('<Draggable />', () => {
  it('should render', async () => {
    const { container, unmount } = render(<Draggable><p>hello</p></Draggable>);
    expect(container).toMatchSnapshot('default');
    fireEvent.dragStart(container.querySelector('p'));
    expect(container).toMatchSnapshot('dragging');
    await sleep(1);
    expect(container).toMatchSnapshot('dragged');
    fireEvent.dragEnd(container.querySelector('p'));
    expect(container).toMatchSnapshot('default again');
    unmount();
  });

  it('should trigger onDragStart props when drag start', () => {
    const onDragStart = vi.fn();
    const onDrag = vi.fn();
    const onDragEnd = vi.fn();
    const { container, unmount } = render(
      <Draggable
        onDrag={onDrag}
        onDragStart={onDragStart}
        onDragEnd={onDragEnd}
      >
        <p>hello</p>
      </Draggable>
    );

    fireEvent.dragStart(container.querySelector('p'));
    expect(onDragStart).toHaveBeenCalled();
    fireEvent.drag(container.querySelector('p'));
    expect(onDrag).toHaveBeenCalled();
    fireEvent.dragEnd(container.querySelector('p'));
    expect(onDragEnd).toHaveBeenCalled();
    unmount();
  });

  it('should not trigger anything when disabled', () => {
    const onDragStart = vi.fn();
    const onDrag = vi.fn();
    const onDragEnd = vi.fn();
    const { container, unmount } = render(
      <Draggable
        disabled={true}
        onDrag={onDrag}
        onDragStart={onDragStart}
        onDragEnd={onDragEnd}
      >
        <p>hello</p>
      </Draggable>
    );

    fireEvent.dragStart(container.querySelector('p'));
    expect(onDragStart).not.toHaveBeenCalled();
    expect(container).toMatchSnapshot('dragging');
    fireEvent.drag(container.querySelector('p'));
    expect(onDrag).not.toHaveBeenCalled();
    fireEvent.dragEnd(container.querySelector('p'));
    expect(onDragEnd).not.toHaveBeenCalled();
    unmount();
  });
});
