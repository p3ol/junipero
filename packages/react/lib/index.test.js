import { useReducer } from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { mockState } from '@junipero/core';
import { useTimeout } from '@junipero/hooks';

import TextField from './TextField';

describe('mockState: inside react', () => {
  it('should fail update the state correctly and return the old ' +
    'state', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch({ value: { ...state.value, content: ['foo'] } });
      }, 1, []);

      const onTextChange = ({ value }) => {
        dispatch({ value: { ...state.value, text: value } });
      };

      const onContentChange = () => {
        dispatch({
          value: { ...state.value, content: [...state.value.content, 'bar'] },
        });
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content but with old text value');

    unmount();
  });

  it('should not fail update the state correctly and return the new ' +
    'state when using mockState with a callback', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch(s => ({ value: { ...s.value, content: ['foo'] } }));
      }, 1, []);

      const onTextChange = ({ value }) => {
        dispatch(s => ({ value: { ...s.value, text: value } }));
      };

      const onContentChange = () => {
        dispatch(s => ({
          value: { ...s.value, content: [...s.value.content, 'bar'] },
        }));
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content AND updated text field valud');

    unmount();
  });
});
