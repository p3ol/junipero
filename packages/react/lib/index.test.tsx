import { useReducer } from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { mockState } from '@junipero/core';
import { useTimeout } from '@junipero/hooks';

import type { FieldContent } from './types';
import TextField from './TextField';

interface State {
  value: {
    text: string;
    content: string[];
  };
}

describe('mockState: inside react', () => {
  it('should correctly update the state and return the new ' +
    'state when using mockState with a callback', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState<State>, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch(s => ({ value: { ...s.value, content: ['foo'] } }));
      }, 1, []);

      const onTextChange = ({ value }: FieldContent) => {
        dispatch(s => ({ value: { ...s.value, text: value } }));
      };

      const onContentChange = () => {
        dispatch(s => ({
          value: { ...s.value, content: [...s.value.content, 'bar'] },
        }));
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content AND updated text field valud');

    unmount();
  });
});
