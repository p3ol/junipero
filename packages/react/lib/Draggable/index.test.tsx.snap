// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Draggable /> should not trigger anything when disabled: dragging 1`] = `
<div>
  <p
    class=""
    draggable="false"
  >
    hello
  </p>
</div>
`;

exports[`<Draggable /> should render: default 1`] = `
<div>
  <p
    class="draggable"
    draggable="true"
  >
    hello
  </p>
</div>
`;

exports[`<Draggable /> should render: default again 1`] = `
<div>
  <p
    class="draggable"
    draggable="true"
  >
    hello
  </p>
</div>
`;

exports[`<Draggable /> should render: dragged 1`] = `
<div>
  <p
    class="dragged draggable"
    draggable="true"
  >
    hello
  </p>
</div>
`;

exports[`<Draggable /> should render: dragging 1`] = `
<div>
  <p
    class="dragging draggable"
    draggable="true"
  >
    hello
  </p>
</div>
`;
